### PR TITLE
Move XFEM Geometric Cut Objects to User Objects

### LINKED ISSUE #9463: Move XFEM Geometric Cut Objects to User Objects
### Description
With an eye to the extensibility of geometric cutting objects in the XFEM module, we aim to move them from being XFEM objects to deriving from GeneralUserObject.

### Rationale
Aside from separating geometric cut objects into their own blocks in input files (allowing for users to more freely add, remove, scale, or translate cuts as they see fit), having cuts as User Objects on their own will allow for automatic determination of points along geometric lines/surfaces in future updates. This will make it much easier to use planned features such as moving interfaces by not requiring a user to manually specify points they would like to move. 

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)
The most public changes are to the syntax of specifying geometric cuts.

### PR DESCRIPTION
**Pull request by jbtompkins**
- Geometric cutting objects now derive from GeneralUserObject (previously XFEM objects)
- All previous geometric cut functionality preserved (multiple cuts in one object, time dependence for 2D lines, etc.)
- Tests updated and pass with new syntax

closes #9463

### Complete each of the following items before creating a PR

- Include any relevant design information for your change
- Follow [Coding Standards](http://mooseframework.org/wiki/CodeStandards/)
- Submit one or more [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/) or modify existing test cases
- Reference or close a specific issue (refs # or closes #)

### CODE REVIEW

File: modules/xfem/include/userobjects/CircleCutUserObject.h [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+template <>
+InputParameters validParams<CircleCutUserObject>();
+
+class CircleCutUserObject : public GeometricCut3DUserObject
+{
+public:
+  CircleCutUserObject(const InputParameters & parameters);
+  ~CircleCutUserObject();
+
+  virtual void initialize(){};
```
Comment (permcody) [inline modules/xfem/include/userobjects/CircleCutUserObject.h]: Use the `override` keyword in all derived objects.

File: modules/xfem/include/userobjects/CircleCutUserObject.h [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+class CircleCutUserObject;
+
+template <>
+InputParameters validParams<CircleCutUserObject>();
+
+class CircleCutUserObject : public GeometricCut3DUserObject
+{
+public:
+  CircleCutUserObject(const InputParameters & parameters);
+  ~CircleCutUserObject();
```
Comment (permcody) [inline modules/xfem/include/userobjects/CircleCutUserObject.h]: Get rid of empty destructors in all derived objects.

File: modules/xfem/include/userobjects/EllipseCutUserObject.h [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+protected:
+  std::vector<Real> _cut_data;
+
+private:
+  std::vector<Point> _vertices;
+  Point _unit_vec1;
+  Point _unit_vec2;
+  Real _long_axis;
+  Real _short_axis;
```
Comment (permcody) [inline modules/xfem/include/userobjects/EllipseCutUserObject.h]: Make all members const when possible if that's your intent. It helps keep the usage clear to the user.

File: modules/xfem/src/userobjects/GeometricCut2DUserObject.C [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+}
+
+bool
+GeometricCut2DUserObject::cutElementByGeometry(const Elem * elem,
+                                               std::vector<CutEdge> & cut_edges,
+                                               Real time) const
+{
+  bool cut_elem = false;
+
+  out << "\n";
```
Comment (permcody) [inline modules/xfem/src/userobjects/GeometricCut2DUserObject.C]: Where is `out` defined? Are you picking this up from the global namespace somehow? `Moose::out`. In any case, if you want to print to the console, use `_console` instead.

File: modules/xfem/src/userobjects/GeometricCut2DUserObject.C [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+  out << "geom_cut_2d _start_times: ";
+  for (unsigned int i = 0; i < _start_times.size(); ++i)
+  {
+    out << _start_times[i];
+    out << ", ";
```
Comment (permcody) [inline modules/xfem/src/userobjects/GeometricCut2DUserObject.C]: This is OK, but you can consolidate this into a single line rather elegantly with the right commas. 
```std::copy(_start_times.begin(), _start_times.end(), infix_ostream_iterator<std::string>(oss, ", "));```
or even better
https://github.com/idaholab/moose/blob/devel/framework/include/utils/Conversion.h#L87-L97

File: modules/xfem/src/userobjects/RectangleCutUserObject.C [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+  params.addClassDescription("Creates a UserObject for planar cuts on 3D meshes for XFEM");
+  // Return the parameters
+  return params;
+}
+
+RectangleCutUserObject::RectangleCutUserObject(const InputParameters & parameters)
+  : GeometricCut3DUserObject(parameters), _cut_data(getParam<std::vector<Real>>("cut_data"))
+{
+  // Throw error if length of cut_data is incorrect
+  if (_cut_data.size() != 12)
```
Comment (permcody) [inline modules/xfem/src/userobjects/RectangleCutUserObject.C]: Magic numbers? Are these documented somewhere?

File: modules/xfem/src/userobjects/RectangleCutUserObject.C [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    Point side_norm = side_tang.cross(_normal);
+    Xfem::normalizePoint(middle2p);
+    Xfem::normalizePoint(side_norm);
+    if (middle2p * side_norm <= 0.0)
+      counter += 1;
+  }
+  if (counter == 4)
+    inside = true;
+  return inside;
+}
```
Comment (permcody) [inline modules/xfem/src/userobjects/RectangleCutUserObject.C]: This class seems to be loaded with magic numbers (constants) Can you use named constants instead where appropriate?

File: modules/xfem/src/userobjects/EllipseCutUserObject.C [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+EllipseCutUserObject::~EllipseCutUserObject() {}
+
+bool
+EllipseCutUserObject::isInsideCutPlane(Point p) const
+{
+  Point ray = p - _center;
+  if (std::abs(ray * _normal) < 1e-6)
+  {
+    double xloc = ray * _unit_vec1;
+    double yloc = ray * _unit_vec2;
```
Comment (permcody) [inline modules/xfem/src/userobjects/EllipseCutUserObject.C]: double -> Real

File: modules/xfem/src/userobjects/LineSegmentCutSetUserObject.C [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+  {
+    Real x0 = (_cut_data[i * 6 + 0] + trans[0]) * scale[0];
+    Real y0 = (_cut_data[i * 6 + 1] + trans[1]) * scale[1];
+    Real x1 = (_cut_data[i * 6 + 2] + trans[0]) * scale[0];
+    Real y1 = (_cut_data[i * 6 + 3] + trans[1]) * scale[1];
+    _cut_line_start_points.push_back(Point(x0, y0, 0.0));
+    _cut_line_end_points.push_back(Point(x1, y1, 0.0));
+
+    _start_times.push_back(_cut_data[i * 6 + 4]);
+    _end_times.push_back(_cut_data[i * 6 + 5]);
```
Comment (permcody) [inline modules/xfem/src/userobjects/LineSegmentCutSetUserObject.C]: More magic numbers, I know you are just doing offsets through your data but it's pretty easy to define these as constants in the class and just use the named versions.

**Review by permcody — DISMISSED** [review]
Just a quick overview of this PR. Not bad.

File: modules/xfem/src/base/XFEM.C [hunk 10]
```diff
@@ -40,17 +38,14 @@ XFEM::XFEM(const InputParameters & params) : XFEMInterface(params), _efa_mesh(Mo
 
 XFEM::~XFEM()
 {
-  for (std::size_t i = 0; i < _geometric_cuts.size(); ++i)
-    delete _geometric_cuts[i];
-
   for (std::map<unique_id_type, XFEMCutElem *>::iterator cemit = _cut_elem_map.begin();
        cemit != _cut_elem_map.end();
        ++cemit)
     delete cemit->second;
```
Comment (dschwen) [inline modules/xfem/src/base/XFEM.C]: We are still using naked new and delete in this subsystem. That's not ideal.

File: modules/xfem/include/base/XFEMCrackGrowthIncrement2DCut.h [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+                                   Real & segment_intersection_fraction);
+
+  Real crossProduct2D(Real ax, Real ay, Real bx, Real by);
+
+  Real _t_start;
+  Real _t_end;
+
+private:
+  const Point _cut_line_start;
+  const Point _cut_line_end;
```
Comment (dschwen) [inline modules/xfem/include/base/XFEMCrackGrowthIncrement2DCut.h]: Just a general remark, you seem to have a lot of data that are pairs of values, so why not make them `std::pair` types?

File: modules/xfem/include/userobjects/CircleCutUserObject.h [hunk 12]
```diff
@@ -0,0 +1,40 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+
+#ifndef CIRCLE_CUT_USEROBJECT_H
+#define CIRCLE_CUT_USEROBJECT_H
```
Comment (dschwen) [inline modules/xfem/include/userobjects/CircleCutUserObject.h]: The convention in framework is _not_ to have underscores in the symbol apart from the one in `_H`

File: modules/xfem/src/base/XFEM.C [hunk 13]
```diff
@@ -1485,7 +1482,7 @@ void
 XFEM::setXFEMCutType(std::string & cut_type)
 {
   _XFEM_cut_type = cut_type;
-}
+}*/
```
Comment (dschwen) [inline modules/xfem/src/base/XFEM.C]: remove commented out suff

File: modules/xfem/src/userobjects/GeometricCut2DUserObject.C [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+        cut_segment = true;
+        segment_intersection_fraction = int_frac;
+      }
+    }
+  }
+  return cut_segment;
+}
+
+Real
+GeometricCut2DUserObject::crossProduct2D(Real ax, Real ay, Real bx, Real by) const
```
Comment (dschwen) [inline modules/xfem/src/userobjects/GeometricCut2DUserObject.C]: would be more elegant if this function took two `const Point &` parameters (to shift the cluttter from the caller side into the function)

Comment (dschwen) [inline modules/xfem/src/userobjects/GeometricCut2DUserObject.C]: Would actually reduce clutter a lot

File: modules/xfem/src/userobjects/GeometricCut3DUserObject.C [hunk 15]
```diff
@@ -86,26 +106,26 @@ XFEMGeometricCut3D::cutElementByGeometry(const Elem * elem,
...
   mooseError("cutFragmentByGeometry not yet implemented for 3D mesh cutting");
   return false;
 }
 
 bool
-XFEMGeometricCut3D::intersectWithEdge(const Point & p1, const Point & p2, Point & pint)
+GeometricCut3DUserObject::intersectWithEdge(const Point & p1, const Point & p2, Point & pint) const
 {
   bool has_intersection = false;
   double plane_point[3] = {_center(0), _center(1), _center(2)};
```
Comment (dschwen) [inline modules/xfem/src/userobjects/GeometricCut3DUserObject.C]: what? `double[3]`? why?

**Review by dschwen — DISMISSED** [review]

File: modules/xfem/src/base/XFEM.C (hunk 10 above)
Comment (nuclear-wizard) [inline modules/xfem/src/base/XFEM.C]: I spoke to Ben and Wen about this, and they said that while this is not an ideal solution, that this and several other naked new and delete instances would be addressed in a future code cleanup pull request for XFEM.

File: modules/xfem/src/userobjects/GeometricCut3DUserObject.C (hunk 15 above)
Comment (nuclear-wizard) [inline modules/xfem/src/userobjects/GeometricCut3DUserObject.C]: @dschwen, @bwspenc, and I spoke about this. The method interfaces with a system with various geometric operations taken into the XFEM module. We deemed changing this system to take moose types beyond the scope of the PR, but it is something that needs to be addressed in a future code cleanup PR.

**Comment by nuclear-wizard** [comment]
Accidentally pushed after rebasing this branch. No need to review this latest commit (changes addressing comments were stashed).

**Comment by jwpeterson** [comment]
@nuclear-wizard you have a merge commit so I don't think you rebased but rather ran `git pull`. You will need to get rid of that (by rebasing) before we can merge this one...

**Comment by jwpeterson** [comment]
@nuclear-wizard looks like this needs to be rebased on upstream/devel one more time. I'm not quite sure how it happened, but you ended up with unrelated commits by Andy Wilkins and Chris Green on your branch. Your PR should only contain the three commits by you. I pushed a rebased version to my fork at jwpeterson/xfem_geometric_cut_userobjects if you are curious what it should look like.

Also your first and last commits have the same subject line, which is a bit confusing (looks like the same commit was accidentally made twice). I'd just change the commit message of the last one to read "Address PR comments by @dschwen and @permcody." and call it good.

**Review by jwpeterson — APPROVED** [review]

**Pull request merged**